### ISSUE TITLE
How to distinguish `\ud808` and `\\ud808`?

### ISSUE THREAD

**Issue by mysteryven** [oxc-project/oxc#2085]
I want to print the String Literal correctly in Codegen, for `'\ud808'` and `'\\ud808'`, it should be different:

```bash
let x = '\\ud808  \ud808' => let x = "\\ud808  \uD808";
```

Our AST node's value is [the same](https://oxc-project.github.io/oxc/playground/?code=3YCAAICZgICAgICAgIC2mcpqppTW%2BJx5VcAGZWOyq7bkGDnoTy6%2Ff3bCAIA%3D). How can I solve it, need to add a `raw` to String Literal?

```bash
"init": {
      "type": "StringLiteral",
      "start": 8,
      "end": 25,
      "value": "\\ud808  \\ud808",
      "raw": "'\\\\ud808  \\ud808'"
 }
```

**Comment by Boshen** [comment]
Can you give me the test case file? It feels like a parser bug, but I'm not sure.

**Comment by mysteryven** [comment]
My brain is a bit of mush, They are from [ESBuild test case](https://github.com/oxc-project/oxc/blob/0e9104477fc94f01b453a19ee01f7f30b2063490/crates/oxc_minifier/tests/esbuild/mod.rs#L207) and [invalid-sccape-sequences.js](https://github.com/tc39/test262/blob/715dd1073bc060f4ee221e2e74770f5728e7b8a0/test/language/expressions/tagged-template/invalid-escape-sequences.js#L48-L51) respectively.

```
 '\\' => {
      let mut cloned_char = chars.clone();
      let next = cloned_char.next();
      let next_next = cloned_char.clone().next();
      if next.is_some_and(|c| matches!(c, 'u' | 'U'))
          && next_next.is_some_and(|c| c.is_ascii_hexdigit())
      {
          p.print_str(b"\\");
          p.print_str(b"u");
          // <https://mathiasbynens.be/notes/javascript-escapes#unicode>
          cloned_char.take(4).for_each(|c| {
              p.print_str(c.to_string().as_bytes());
          });
          for _ in 0..5 {
              chars.next();
          }
      } else {
          p.print_str(b"\\\\");
      }
  }
```


After passing the ESBuild test case, I converted `\\u0g` to '\u0g' accidentally, which not a valid Unicode, seems I need to know whether `\` is being escaped, our AST not show this.

**Comment by Boshen** [comment]
For `test("let x = '\\uD808'", "let x = \"\\uD808\";");`, the right hand side may be incorrect due to escaping rules when we ported the test case over, which means we may need to fix the right hand side value for this one.

The easiest way to fix this is copy and paste the value in node.js, and then copy it out again into Rust, so it's actual value is preserved.
---

For 

```
(strs => {
  assert.sameValue(strs[0], undefined, 'Cooked template value should be undefined for illegal escape sequences');
  assert.sameValue(strs.raw[0], '\\u00g');
})`\u00g`;
```

This is entirely different, this is accessing the cooked value, it has different escaping rules for tagged vs untagged template (you need to check this, I may be wrong).

**Comment by mysteryven** [comment]
There are no related test cases, if the string value is only a high surrogate:

```
let x == '\ud808'
```

the [playground](https://oxc-project.github.io/oxc/playground/?code=3YCAAICQgICAgICAgIC2mcpqppTW%2BJx5VHQktxlcQwOCrWt%2FfZtAgA%3D%3D) will be:

```
"init": {
    "type": "StringLiteral",
    "start": 8,
    "end": 16,
    "value": "\\ud808"
  },
```


the Codegen output will be:

```
let x = '\\ud808';
```

How can I print this right? :)

**Comment by Boshen** [comment]
Close for house keeping. We'll deal with this problem when we focus on the codegen.
